file/shard.go

```diff
@@ -110,23 +194,16 @@ func (s *shardNodeReader) Seek(offset int64, whence int) (int64, error) {
 
 func (s *shardNodeFile) length() int64 {
 	// see if we have size specified in the unixfs data. errors fall back to length from links
-	nodeData, err := s.substrate.LookupByString("Data")
-	if err != nil {
-		return s.lengthFromLinks()
-	}
-	nodeDataBytes, err := nodeData.AsBytes()
+	nodeData, err := s.unpack()
 	if err != nil {
```

Comment:
similar to my note suggestion above, we can have the `nodeData==nil` case even without an error if `s.unpack()` has already been called. Unless we want to also cache `retError` on `shardNodeFile` along with `metadata`?

```suggestion
	if err != nil || nodeData == nil {
```